vision/conv_mnist/conv_mnist.jl

```diff
@@ -1,240 +1,215 @@
+
+#===== TRAINING =====#
+
+# Let's collect some hyper-parameters in a NamedTuple, just to write them in one place.
+# Global variables are fine -- we won't access this from inside any fast loops.
 
-# * **Loads MNIST dataset:** Loads the train and test set tensors. The shape of the train data is `28x28x60000` and the test data is `28x28x10000`. 
-# * **Reshapes the train and test data:**  Notice that we reshape the data so that we can pass it as arguments for the input layer of the model.
-# * **One-hot encodes the train and test labels:** Creates a batch of one-hot vectors so we can pass the labels of the data as arguments for the loss function. For this example, we use the [logitcrossentropy](https://fluxml.ai/Flux.jl/stable/models/losses/#Flux.Losses.logitcrossentropy) function and it expects data to be one-hot encoded. 
-# * **Creates mini-batches of data:** Creates two DataLoader objects (train and test) that handle data mini-batches of size `128 ` (as defined above). We create these two objects so that we can pass the entire data set through the loss function at once when training our model. Also, it shuffles the data points during each iteration (`shuffle=true`).
+settings = (;
+    eta = 3e-4,     # learning rate
+    lambda = 1e-2,  # for weight decay
+    batchsize = 128,
+    epochs = 10,
+)
+train_log = []
 
-# ## Model
+# Initialise the storage needed for the optimiser:
 
-# We create the LeNet5 "constructor". It uses Flux's built-in [Convolutional and pooling layers](https://fluxml.ai/Flux.jl/stable/models/layers/#Convolution-and-Pooling-Layers):
+opt_rule = OptimiserChain(WeightDecay(settings.lambda), Adam(settings.eta))
+opt_state = Flux.setup(opt_rule, lenet);
 
+for epoch in 1:settings.epochs
+    # @time will show a much longer time for the first epoch, due to compilation
+    @time for (x,y) in loader(batchsize=settings.batchsize)
+        grads = Flux.gradient(m -> Flux.logitcrossentropy(m(x), y), lenet)
+        Flux.update!(opt_state, lenet, grads[1])
+    end
 
-function LeNet5(; imgsize=(28,28,1), nclasses=10) 
-    out_conv_size = (imgsize[1]÷4 - 3, imgsize[2]÷4 - 3, 16)
-    
-    return Chain(
-            Conv((5, 5), imgsize[end]=>6, relu),
-            MaxPool((2, 2)),
-            Conv((5, 5), 6=>16, relu),
-            MaxPool((2, 2)),
-            flatten,
-            Dense(prod(out_conv_size), 120, relu), 
-            Dense(120, 84, relu), 
-            Dense(84, nclasses)
-          )
+    # Logging & saving, but not on every epoch
+    if epoch % 2 == 1
+        loss, acc, _ = loss_and_accuracy(lenet)
+        test_loss, test_acc, _ = loss_and_accuracy(lenet, test_data)
+        @info "logging:" epoch acc test_acc
+        nt = (; epoch, loss, acc, test_loss, test_acc)  # make a NamedTuple
+        push!(train_log, nt)
+    end
+    if epoch % 5 == 0
+        filename = joinpath("runs", "lenet.jld2")
+        JLD2.jldsave(filename; lenet_state = Flux.state(lenet |> cpu))
```

Comment:
I changed this to save using `state` and JLD2.

Unlike BSON, what this won't do is make the folder to write into. Not sure what's best. Could make it manually. Or just change gitignore to have `*.jld2`?

Reply:
[b32a41c](https://github.com/FluxML/model-zoo/pull/394/commits/b32a41c35b943bf37cfecad26375582d5c7c9074) makes the folder by hand